compiler/rustc_transmute/src/maybe_transmutable.rs

```diff
@@ -0,0 +1,248 @@
+        scope,
+        assume_alignment,
+        assume_lifetimes,
+        assume_validity,
+        assume_visibility,
+        within_references,
+        tcx: tcx.tcx(),
+        cache: Map::default(),
+    }
+    .answer(src_nfa.start, dst_nfa.start)
+}
+
+struct MaybeTransmutableQuery<'tcx, 'nfa, C>
+where
+    C: HasTyCtxt<'tcx>,
+{
+    src_nfa: &'nfa Nfa<'tcx>,
+    dst_nfa: &'nfa Nfa<'tcx>,
+    scope: Ty<'tcx>,
+    assume_alignment: bool,
+    assume_lifetimes: bool,
+    assume_validity: bool,
+    assume_visibility: bool,
+    within_references: bool,
+    tcx: C,
+    cache: Map<(State, State), Answer<'tcx>>,
+}
+
+impl<'tcx, 'nfa, C> MaybeTransmutableQuery<'tcx, 'nfa, C>
+where
+    C: HasTyCtxt<'tcx>,
+{
+    pub fn answer(&mut self, src_state: State, dst_state: State) -> Answer<'tcx> {
+        let empty_map = Map::default();
+        let empty_set = Set::default();
+        if dst_state == self.dst_nfa.accepting {
+            // truncation: `size_of(Src) >= size_of(Dst)`
+            Answer::Yes
+        } else if (src_state == self.src_nfa.accepting) && !self.within_references {
+            // extension: `size_of(Src) < size_of(Dst)`
+            // the remaining bytes of Dst must accept `Uninit`
+            let dst_state_primes = self
+                .dst_nfa
+                .edges_from(dst_state)
+                .unwrap_or(&empty_map)
+                .get(&Transition::Byte(Byte::Uninit))
+                .unwrap_or(&empty_set);
+
+            there_exists(dst_state_primes, |&dst_state_prime| {
+                self.answer_cached(src_state, dst_state_prime)
+            })
+        } else {
+            let src_quantification = if self.assume_validity {
+                // if the compiler may assume that the programmer is doing additional validity checks,
+                // (e.g.: that `src != 3u8` when the destination type is `bool`)
+                // then there must exist at least one transition out of `src_state` such that the transmute is viable...
+                there_exists
+            } else {
+                // if the compiler cannot assume that the programmer is doing additional validity checks,
+                // then for all transitions out of `src_state`, such that the transmute is viable...
+                // then there must exist at least one transition out of `src_state` such that the transmute is viable...
+                for_all
+            };
+            src_quantification(
+                self.src_nfa.edges_from(src_state).unwrap_or(&empty_map),
+                |(&src_transition, src_state_primes)| {
+                    let dst_quantification = if self.assume_validity {
+                        // for some successive src state reached via `src_transition`...
+                        there_exists
+                    } else {
+                        // for all successive src states reached via `src_transition`...
+                        for_all
+                    };
```

Comment:
Is this qualification necessary?
As far as I understand, the condition should be:
- (assume.validity==false): all paths of Src must be accepted by some path of Dst
- (assume.validity==true): some paths of Src must be accepted by some path of Dst
Is my understanding wrong?

Reply:
The understanding was indeed wrong, hiding the comment

Reply:
Yeah, this seems suspicious.